### PR TITLE
RHCLOUD-46516 - Enhance testing guidelines with TDD and no-mocks philosophy

### PR DESCRIPTION
**Pull request by snehagunta**
## Summary

This PR enhances the agent testing guidelines established in #1292 by adding comprehensive testing philosophy emphasizing TDD and a no-mocks approach. This continues the RHCLOUD-46516 work to improve AI agent coding practices.

**Key additions:**
- **TDD philosophy**: Establish test-driven development as the preferred approach for all features and bug fixes
- **No-mocks approach**: Define clear preference order (real → fake → stub) and discourage method-verifying mocks
- **Hermetic testing**: Guide agents to use testcontainers when external dependencies are truly needed
- **Contract testing**: Document the pattern for validating fakes against real implementations
- **Expanded documentation**: Catalog all existing fake implementations with their purposes

**Changes:**
- Add "Testing Philosophy" section to testing guidelines with TDD and no-mocks guidance
- Document preference order for test doubles with examples
- Add contract testing pattern with code examples
- Expand fake implementations table to include all repository fakes
- Clarify when to use SQLite vs fake repositories
- Update anti-patterns to discourage method-verifying mocks
- Update AGENTS.md agent responsibilities to require TDD approach

## Test plan
- [x] Review testing-guidelines.md for accuracy and completeness
- [x] Verify AGENTS.md properly links to testing guidelines
- [x] Confirm all fake implementations mentioned exist in codebase
- [x] Validate contract testing example matches actual patterns in `resource_repository_test.go`

## Related
- Part of RHCLOUD-46516 epic for using agents to control coding practices
- Builds on initial agent readiness work from #1292
- Supports layered architecture refactoring epic (RHCLOUD-45xxx series)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Documentation**
  * Updated testing guidelines to establish test-driven development (TDD) as the default workflow for all code contributions, ensuring features are validated from inception
  * Introduced refined testing strategies emphasizing contract testing and improved test design practices
  * Enhanced development standards to ensure more reliable and maintainable code delivery

### CODE REVIEW

File: docs/testing-guidelines.md [hunk 1]
```diff
@@ -1,5 +1,23 @@
 # Testing Guidelines for Inventory API
 
+## Testing Philosophy
+
+**Prefer test-driven development (TDD).** Write tests first to define the expected behavior, then write the implementation to make them pass. TDD produces cleaner interfaces, better coverage, and catches design problems early. When adding a new feature or fixing a bug, start by writing a failing test that captures the requirement, then implement the minimum code to pass it.
```
Comment (merlante) [inline docs/testing-guidelines.md:5]: Is this for agents or for humans? :) The reason I ask is that I don't know if agents being instructed to follow a TDD approach would make them better or worse.

Comment (tylercreller) [reply docs/testing-guidelines.md:5]: also curious about this, I have not tried a TDD flow yet with agents

Comment (snehagunta) [reply docs/testing-guidelines.md:5]: I am trying TDD as a human working with an agent. With the refactoring epic, I give it the story context and ask it to tell me what tests are missing or need to be written before I implement the story. It does a good job of telling me what tests to write. Although the tests themselves are not necessarily good because it follows the patterns that exist around it, which are not consistent or always good. 

As for the agent doing it automatically, I am experimenting here. Maybe it does, maybe it does not, I am not sure. This was generated by claude fwiw and it said to me that it will look up the guidelines, so I want to see how it works. I am open to other ideas too.

Comment (Rajagopalan-Ranganathan) [reply docs/testing-guidelines.md:5]: +1 I have always followed TDD or testing first approach, especially for refactors, to capture the existing behavior. Such as we dont add any regression.
For new pieces of s/w we want, agents might and will go full blown, if we dont constraint them, Tests are a good guard rail for enforcing things, "doing the minimal" principle. I would like to try this out, and see where it can improved.

**Review by Rajagopalan-Ranganathan — APPROVED** [review]

**Review by tylercreller — APPROVED** [review]
